metagraph/src/annotation/binary_matrix/row_diff/row_diff.cpp

```diff
@@ -42,6 +47,60 @@ BinaryMatrix::SetBitPositions RowDiff::get_row(Row row) const {
     return result;
 }
 
+std::vector<BinaryMatrix::SetBitPositions>
+RowDiff::get_rows(const std::vector<Row> &row_ids) const {
+    std::vector<SetBitPositions> rows(row_ids.size());
+
+    tsl::ordered_map<graph::boss::BOSS::edge_index, SetBitPositions> cached_rows;
+    cached_rows.reserve(rows.size() * 2);
+
+    for (size_t i = 0; i < row_ids.size(); ++i) {
+        // reset the cache if it's too large
+        if (cached_rows.size() > MIN_ROW_CACHE_CAPACITY
+                                    + row_ids.size() * ROW_CACHE_CAPACITY_FACTOR)
+            cached_rows.clear();
```

Comment:
it seems wasteful to clear the cache like this, could we not use one of the caches from the external library for this?

Reply:
That would take more code and it's unlikely that it would work any faster.
if we assume the locality, like `row_ids[i + 1]` is faster to query if the results from `row_ids[i]` are cached, both approaches work equally well. If there is no locality, on the other hand, no caching strategy will work.

So if we pass `row_ids` in a nice order, it doesn't matter how large is the cache and how we reinitialize it.

Reply:
but maybe it makes sense to use a cache lib, to get the thread-safety easily, so we can run this loop with multiple threads. I'll just need to check if that doesn't make it slower with 1 thread.

Reply:
I figured it's better to assume the batch is small enough so we don't need to reset the buffer.

If a batch is too large and the indexes are shuffled, the cache will bring very little benefit.

So by never resetting the cache, we make sure that the entire batch is processed in the most efficient way.